Tests/YSwiftTests/YUndoManagerTests.swift

```diff
@@ -0,0 +1,130 @@
+import Combine
+import XCTest
+@testable import YSwift
+
+class TestMetadata {
+    public let value: String
+    init(_ value: String) {
+        self.value = value
+    }
+}
+
+func exchangeUpdates(_ document1: YDocument, _ document2: YDocument) {
+    let clock1 = document1.transactSync { txn in
+        txn.transactionStateVector()
+    }
+    let clock2 = document2.transactSync { txn in
+        txn.transactionStateVector()
+    }
+    let update1 = document1.transactSync { txn in
+        document1.diff(txn: txn, from: clock2)
+    }
+    let update2 = document2.transactSync { txn in
+        document2.diff(txn: txn, from: clock1)
+    }
+    document1.transactSync { txn in
+        try! txn.transactionApplyUpdate(update: update2)
+    }
+    document2.transactSync { txn in
+        try! txn.transactionApplyUpdate(update: update1)
+    }
+}
+
+class YUndoManagerTests: XCTestCase {
+    var document: YDocument!
+    var text: YText!
+    var manager: YUndoManager<TestMetadata>!
+    
+    override func setUp() {
+        document = YDocument()
+        text = document.getOrCreateText(named: "test")
+        manager = document.undoManager(trackedRefs: [text])
+    }
+    
+    override func tearDown() {
+        document = nil
+        text = nil
+        manager = nil
+    }
+    
+    
+    func test_undoTextBasic() throws {
+        text.insert("1221", at: 0)
+        manager.wrap()
+        text.insert("3", at: 2)
+        manager.wrap()
+        text.insert("3", at: 3)
+        manager.wrap()
+        
+        XCTAssertEqual(text.getString(), "123321")
+        
+        XCTAssert (try manager.undo())
+        XCTAssertEqual(text.getString(), "12321")
+        
+        XCTAssert (try manager.undo())
+        XCTAssertEqual(text.getString(), "1221")
+        
+        text.insert("3", at: 2)
+        XCTAssertEqual(text.getString(), "12321")
+    }
+    
+    func test_undoEvents() throws {
+        var received = TestMetadata("")
+        let on_added = manager.observeAdded({ (e, metadata) -> TestMetadata? in
+            let m = metadata ?? TestMetadata("")
+            return TestMetadata(m.value + "A")
+        })
+        let on_popped = manager.observePopped({ (e, metadata) -> TestMetadata? in
+            received = metadata!
+            return metadata
+        })
+        
+        text.insert("abc", at: 0)
+        
+        XCTAssert (try manager.undo())
+        XCTAssertEqual(received.value, "A")
+        
+        XCTAssert (try manager.redo())
+        XCTAssertEqual(received.value, "A")
+    }
+    
+    func test_undoDifferentOrigins() throws {
+        let localOrigin = Origin("local")
+        
+        let remoteDocument = YDocument()
+        let remoteText = document.getOrCreateText(named: "test")
+        manager.addOrigin(localOrigin) // only track transaction from local origin
+        
+        // create some changes locally
+        document.transactSync(origin: localOrigin) { txn in
```

Comment:
For some reason when I try to call transactSync with origin provided, `swift test` exits with error code 5. Any ideas? @heckj @nugmanoff

Reply:
It's hitting a dispatchPrecondition - we have an assertion in there to verify that all calls that manipulate transactions are on a specific DispatchQueue (`transactionQueue`) that's defined in `YDocument`.

Screenshot from Xcode shows the call stack on the left:
<img width="1398" alt="Screenshot 2024-03-03 at 10 50 34 AM" src="https://github.com/y-crdt/yswift/assets/43388/4ac97496-1022-428c-9cc0-f6a0512c3c09">

Reply:
The insert in the text is attempting to set up it's own transaction, expecting to be called externally from `YDocument`:
```swift
    public func insert(
        _ text: String,
        at index: UInt32,
        in transaction: YrsTransaction? = nil
    ) {
        withTransaction(transaction) { txn in
            self._text.insert(tx: txn, index: index, chunk: text)
        }
    }
```

It's the interior call setting up the transaction again that's breaking things. Right now the boundary is all the public methods on `YDocument` are wrapped with their own transaction interface, which is ghetto - but works for the simple bits. For your test, you could manipulate the internal bits directly:
`self._text.insert(tx: txn, index: index, chunk: text)` - but it might be better to set up UndoManager within the context of Document, so that calls from it can work directly on the document, and then we can verify the "who owns this" boundary at the YDocument level.

The whole process of using a DispatchQueue here is functional, but going forward we'll do better to embrace Swift's concurrency (very akin to Rust's borrow checker setup) and leverage async/await and actors protect the internal interface to Rust. UniFFI 0.26 recently added the ability to pass async function calls cross the Rust/Swift boundary, if that's of potential future use as well - I haven't done anything with it, just noted it when I updated the dependency.